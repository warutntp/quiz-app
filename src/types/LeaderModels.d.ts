export module LeaderModels {
  export interface LeaderModel {
    name: string;
    score: number;
  }
}
