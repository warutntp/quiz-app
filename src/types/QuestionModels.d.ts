export module QuestionModels {
  export interface QuestionModel {
    question: string;
    answers: string[];
    correct: string;
  }
}
