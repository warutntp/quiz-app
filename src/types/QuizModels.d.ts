export module QuizModels {
  export interface QuizModel {
    question: string;
    answers: string[];
    correct: string;
  }
}
