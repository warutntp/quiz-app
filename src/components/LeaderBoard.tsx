import React from "react";
import { LeaderModels } from "../types/LeaderModels";

interface LeaderBoardProps {
  scores: LeaderModels.LeaderModel[];
}

const LeaderBoard: React.FC<LeaderBoardProps> = ({ scores }) => {
  const sortedScores = scores.slice().sort((a, b) => b.score - a.score);

  return (
    <div className="p-4 border rounded-lg shadow-md bg-white mt-4 max-h-[500px] overflow-y-auto">
      <h2 className="text-lg font-bold mb-4">LeaderBoard</h2>
      <ul>
        {sortedScores.map((entry, index) => (
          <li key={index} className="flex justify-between p-2 border-b">
            <span>{entry.name}</span>
            <span>{entry.score}</span>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default LeaderBoard;
