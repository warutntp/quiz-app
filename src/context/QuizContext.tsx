import React, {
  createContext,
  useReducer,
  ReactNode,
  Dispatch,
  useEffect,
  useMemo,
} from "react";
import { QuizModels } from "../types/QuizModels";
import { LeaderModels } from "../types/LeaderModels";
import { RandomData } from "../utils/RandomData";
import { QuizData } from "../data/QuizData";
import {
  loadLeaderDataFromLocalStorage,
  saveLeaderDataToLocalStorage,
} from "../utils/LocalStorage";

type State = {
  currentQuestion: QuizModels.QuizModel | null;
  leaderBoard: LeaderModels.LeaderModel[];
  name: string;
  score: number;
  isQuizActive: boolean;
  questionIndex: number;
  totalQuestions: QuizModels.QuizModel[];
  answers: string[];
};

const initialState: State = {
  currentQuestion: null,
  leaderBoard: [],
  name: "",
  score: 0,
  isQuizActive: false,
  questionIndex: 0,
  totalQuestions: [],
  answers: [],
};

type Action =
  | { type: "SET_NAME"; payload: string }
  | { type: "START_QUIZ" }
  | { type: "ANSWER_QUESTION"; payload: string }
  | { type: "NEXT_QUESTION" }
  | { type: "PREVIOUS_QUESTION" }
  | { type: "SUBMIT_QUIZ" }
  | { type: "LOAD_LEADERBOARD"; payload: LeaderModels.LeaderModel[] };

const quizReducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "SET_NAME":
      return { ...state, name: action.payload };
    case "START_QUIZ": {
      const shuffled = RandomData([...QuizData], 20);
      return {
        ...state,
        isQuizActive: true,
        totalQuestions: shuffled,
        currentQuestion: shuffled[0],
        questionIndex: 0,
        answers: Array(shuffled.length).fill(null),
      };
    }
    case "ANSWER_QUESTION": {
      const newAnswers = [...state.answers];
      newAnswers[state.questionIndex] = action.payload;
      return { ...state, answers: newAnswers };
    }
    case "NEXT_QUESTION":
      if (state.questionIndex < state.totalQuestions.length - 1) {
        return {
          ...state,
          questionIndex: state.questionIndex + 1,
          currentQuestion: state.totalQuestions[state.questionIndex + 1],
        };
      }
      return state;
    case "PREVIOUS_QUESTION":
      if (state.questionIndex > 0) {
        return {
          ...state,
          questionIndex: state.questionIndex - 1,
          currentQuestion: state.totalQuestions[state.questionIndex - 1],
        };
      }
      return state;
    case "SUBMIT_QUIZ": {
      const newScore = state.answers.reduce((acc, answer, idx) => {
        if (answer === state.totalQuestions[idx].correct) {
          return acc + 1;
        }
        return acc;
      }, 0);

      const newLeaderBoard = [
        ...state.leaderBoard,
        { name: state.name, score: newScore },
      ];
      saveLeaderDataToLocalStorage(newLeaderBoard);

      return {
        ...state,
        score: newScore,
        isQuizActive: false,
        leaderBoard: newLeaderBoard,
      };
    }
    case "LOAD_LEADERBOARD":
      return { ...state, leaderBoard: action.payload };
    default:
      return state;
  }
};

interface QuizContextProps {
  state: State;
  dispatch: Dispatch<Action>;
}

const QuizContext = createContext<QuizContextProps>(undefined as any);

const QuizProvider: React.FC<{ children?: ReactNode }> = ({ children }) => {
  const [state, dispatch] = useReducer(quizReducer, initialState);

  useEffect(() => {
    const savedLeaderBoard = loadLeaderDataFromLocalStorage();
    dispatch({ type: "LOAD_LEADERBOARD", payload: savedLeaderBoard });
  }, [dispatch]);

  const contextValue = useMemo(() => ({ state, dispatch }), [state, dispatch]);

  return (
    <QuizContext.Provider value={contextValue}>{children}</QuizContext.Provider>
  );
};

export { QuizContext, QuizProvider };
