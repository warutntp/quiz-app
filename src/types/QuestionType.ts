export interface QuestionType {
  question: string;
  answers: string[];
  correct: string;
}
