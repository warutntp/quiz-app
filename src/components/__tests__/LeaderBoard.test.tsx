import { render, screen } from "@testing-library/react";
import LeaderBoard from "../LeaderBoard";
import "@testing-library/jest-dom/extend-expect";
import { LeaderModels } from "../../types/LeaderModels";
import { QuizProvider } from "../../context/QuizContext";
const mockScores: LeaderModels.LeaderModel[] = [
  { name: "John", score: 100 },
  { name: "Jane", score: 90 },
  { name: "Doe", score: 80 },
];

test("LeaderBoard", () => {
  const { container } = render(
    <QuizProvider>
      <LeaderBoard scores={mockScores} />
    </QuizProvider>
  );
  expect(container).toMatchSnapshot();
});

test("renders leaderboard correctly", () => {
  render(
    <QuizProvider>
      <LeaderBoard scores={mockScores} />
    </QuizProvider>
  );

  expect(screen.getByText("LeaderBoard")).toBeInTheDocument();
  expect(screen.getByText("John")).toBeInTheDocument();
  expect(screen.getByText("100")).toBeInTheDocument();
  expect(screen.getByText("Jane")).toBeInTheDocument();
  expect(screen.getByText("90")).toBeInTheDocument();
  expect(screen.getByText("Doe")).toBeInTheDocument();
  expect(screen.getByText("80")).toBeInTheDocument();
});
