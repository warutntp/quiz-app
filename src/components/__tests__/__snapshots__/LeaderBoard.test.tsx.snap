// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderBoard 1`] = `
<div>
  <div
    class="p-4 border rounded-lg shadow-md bg-white mt-4 overflow-y-auto max-h-[600px]"
  >
    <h2
      class="text-lg font-bold mb-4"
    >
      LeaderBoard
    </h2>
    <ul>
      <li
        class="flex justify-between p-2 border-b"
      >
        <span>
          John
        </span>
        <span>
          100
        </span>
      </li>
      <li
        class="flex justify-between p-2 border-b"
      >
        <span>
          Jane
        </span>
        <span>
          90
        </span>
      </li>
      <li
        class="flex justify-between p-2 border-b"
      >
        <span>
          Doe
        </span>
        <span>
          80
        </span>
      </li>
    </ul>
  </div>
</div>
`;
